autoload/go/def.vim

```diff
@@ -114,11 +135,122 @@ function! s:godefJump(out, mode)
 			endif
 		endif
 
-		" jump to file now
-		sil ll 1
-		normal zz
+        " Remove anything newer than the current position, just like basic
+        " vim tag support
+        if w:go_stack_level == 0
+            let w:go_stack = []
+        else
+            let w:go_stack = w:go_stack[0:w:go_stack_level-1]
+        endif
+
+		" increment the stack counter
+		let w:go_stack_level = w:go_stack_level + 1
 
-		let &switchbuf = old_switchbuf
+		" push it on to the jumpstack
+        call add(w:go_stack,
+            \{'line': line("."), 'col': col("."),
+            \'file': expand('%:p'), 'ident': a:out[1]})
+
+		" jump to file now
+        call s:goToFileLocation(location)
 	end
-	let &errorformat = old_errorformat
+endfunction
+
+function! go#def#StackPrint()
+    if len(w:go_stack) == 0
+        echohl ErrorMsg
+        echo "godef stack empty"
+        echohl None
+        return
+    endif
+    let i = 0
+    while i < len(w:go_stack)
+        let entry = w:go_stack[i]
+        if i == w:go_stack_level
+            echon "> "
+        else
+            echon "  "
+        endif
+        echon i+1 . " "
+        echohl Directory
+        echon entry["file"]
+        echohl None
+        echon "|"
+        echohl LineNr
+        echon entry["line"] . " col " . entry["col"]
+        echohl None
+        echon "|"
+        echon entry["ident"] . "\n"
+        let i += 1
+    endwhile
+    if w:go_stack_level == i
+        echo ">"
+    endif
+endfunction
+
+function! go#def#StackPop(numPop)
+    let newLevel = str2nr(w:go_stack_level) - str2nr(a:numPop)
+    if newLevel < 0
+        echohl ErrorMsg
+        echo "at bottom of godef stack"
+        echohl None
+        let newLevel = 0
+    endif
+    if w:go_stack_level > 0
+        let w:go_stack_level = newLevel
+        let target = w:go_stack[w:go_stack_level]
+
+        " jump
+        call s:goToFileLocation(target["file"], target["line"], target["col"])
+    endif
+endfunction
+
+function! go#def#StackJump(...)
+    if len(w:go_stack) == 0
+        echohl ErrorMsg
+        echo "godef stack empty"
+        echohl None
+        return
+    endif
+	if !len(a:000)
+        " Display interactive stack
+        call go#def#StackPrint()
+        echon "\n"
+        let jumpTarget=input("Type number and <Enter> (empty cancels): ")
+	else
+		let jumpTarget= a:1
+	endif
+
+    let jumpTarget=str2nr(jumpTarget) - 1
+    if jumpTarget >= 0 && jumpTarget < len(w:go_stack)
+        let w:go_stack_level = jumpTarget
+        let target = w:go_stack[w:go_stack_level]
+
+        " jump
+        call s:goToFileLocation(target["file"], target["line"], target["col"])
+    else
+        echohl ErrorMsg
+        echo "invalid godef stack location"
+        echohl None
+    endif
+endfunction
+
+function! s:goToFileLocation(...)
```

Comment:
The function signature should be explicit about the given arguments. Like: `s:goToFileLocation(a:file, a:line, a:col)`